codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithyjava/modeled/ModeledShapeValue.java

```diff
@@ -0,0 +1,180 @@
+    private static class ValueNodeVisitor implements NodeVisitor<CodeBlock> {
+        private final JavaLibrary subject;
+        private final Shape shape;
+
+        // https://smithy.io/2.0/spec/model.html#trait-node-values says the string value should be base64 encoded,
+        // but both https://smithy.io/2.0/additional-specs/http-protocol-compliance-tests.html
+        // and https://smithy.io/2.0/additional-specs/smoke-tests.html
+        // say "binary data MUST be defined using values that can be represented in plain text".
+        // Given this requirement is driven by making it easier to assert that data is automatically
+        // base64 encoded in http protocol compliance tests,
+        // which isn't as relevant for polymorph libraries,
+        // we support either interpretation depending on this flag.
+        // That way we can implement @smokeTests as specified
+        // but use modeled node generation for other purposes that will want to
+        // support arbitrary binary data.
+        private final boolean base64BlobStrings;
+
+        private ValueNodeVisitor(JavaLibrary subject, boolean base64BlobStrings, Shape shape) {
+            this.subject = subject;
+            this.shape = shape;
+            this.base64BlobStrings = base64BlobStrings;
+        }
+
+        private ValueNodeVisitor visitorForShape(Shape shape) {
+            return new ValueNodeVisitor(subject, base64BlobStrings, shape);
+        }
+
+        @Override
+        public CodeBlock objectNode(ObjectNode node) {
+            return switch (shape.getType()) {
+                case STRUCTURE, UNION -> structureOrUnionShape(shape, node);
+                case MAP -> mapShape((MapShape) shape, node);
+                case DOCUMENT -> documentShape((DocumentShape) shape, node);
+                default -> throw new CodegenException("unexpected input shape: " + shape.getType());
+            };
+        }
+
+        private CodeBlock structureOrUnionShape(Shape shape, ObjectNode node) {
+            // The builder interfaces for structures and unions are identical,
+            // the only difference is the constraint on only setting one member.
+            if (shape instanceof UnionShape && node.getMembers().size() != 1) {
+                throw new IllegalArgumentException("Union shapes require an object node with exactly one member");
+            }
+
+            final TypeName inputType = subject.nativeNameResolver.typeForShape(shape.getId());
+            CodeBlock.Builder result = CodeBlock.builder().add("$T.builder()", inputType);
+
+            for (Map.Entry<String, Node> entry : node.getStringMap().entrySet()) {
+                String memberName = entry.getKey();
+                Node memberValue = entry.getValue();
+
+                MemberShape memberShape = shape.getAllMembers().get(memberName);
+                Shape targetShape = subject.model.expectShape(memberShape.getTarget());
```

Comment:
Ah great question. You're talking about how to handle `@references` shapes in other words. 

In practical terms `smokeTests` isn't going to be able to handle that, because resources aren't expressible as a simple Node literal. You need at least a sequence of calls: one call to say `Create***Keyring` to create a reference, and then another call to bind that as input to another call. I can imagine having a more beefed-up version of specifying a test that supports sequences, but it gets a lot more complicated real fast, and at that point I think you're better off writing a Dafny test instead so you're using a proper language.

But this does make me realize that I should raise an explicit error if the shape has `@references` on it, in case someone tries to misuse this by accident.

Reply:
This is a good answer, but it raises another question.

Can you, with documentation only, state what the scope of Smoke Test is?

It sounds like it is for Testing Smithy-Dafny ONLY, at least today.
As compared to testing Smithy-Dafny's MPL or DB-ESDK generated artifacts.
(Because those would need `@references` support.)

This a very reasonable (and valuable!) scope, 
but we need to tell people.

Reply:
Ah I had it in my todos to consider adding something under `designs` for this, but I think the easiest place is to add a case under https://github.com/smithy-lang/smithy-dafny/tree/main-1.x/TestModels#testing-patterns, so I'll do that.

I'm also going to note more visibility that the way we have mapped shapes to Dafny types makes it impossible to implement `smokeTests` for Dafny, which is fine but not obvious.

Reply:
`designs` is buried from users, as is Testing-Patterns.

The `output-java-test` argument you added only supports Smoke-Tests, correct?

That is GREAT spot to tell users this is for internal testing only.

Adding some comments to the implementations should inform any Smithy-Dafny contributors.

And this really is only for internal testing,
as `@references` is a prominent Smithy-Dafny feature.

Reply:
> The output-java-test argument you added only supports Smoke-Tests, correct?

At the moment yes, but there are other Smithy traits that could generate tests in the future (existing ones like those in https://smithy.io/2.0/additional-specs/http-protocol-compliance-tests.html or future ones we might invent ourselves) and they wouldn't necessarily be internal-only, so it feels a little wrong to give a warning there.

I think it would be better to create a direct warning on the use of `smokeTests` outside of the test models. There might be a clean way to do that based on `metadata suppressions`, let me experiment.